book/src/persistent-account-storage.md

```diff
@@ -0,0 +1,120 @@
+# Persistent Account Storage
+
+The set of Accounts represent the current computed state of all the transactions
+that have been processed by a fullnode.  Each fullnode needs to maintain this
+entire set.  Each block that is proposed by the network represents a change to
+this set, and since each block is a potential rollback point the changes need to
+be reversible.
+
+Persistent storage like NVMEs are 20 to 40 times cheaper then DDR.  The problem
+with persistent storage is that write and read performance is much slower then
+DDR, and care must be taken in how data is read or written to.  Both reads and
+writes can be split between multiple storage drives and accessed in parallel.
+This design proposes a data structure that allows for concurrent reads and
+concurrent writes of storage.   Writes are optimized by using an AppendVec data
+structure, which allows a single writer to append while allowing access to many
+concurrent readers.  The accounts index maintains a pointer to a spot where the
+account was appended to for every fork, thus removing the need for explicit
+check-pointing of state.
+
+# AppendVec
+
+AppendVec is a data structure that allows for random reads concurrent with a
+single append only writer.  Grow, or resizing the capacity of the AppendVec
+requires exclusive access.  This is implemented with an atomic `len`, which is
+updated at the end of a completed append.
+
+The underlying memory for an AppendVec is a memory mapped file.  Memory mapped
+files allow for fast random access, and paging is handled by the OS.
+
+# Account Index
+
+The account index is designed to support a single index for all the currently
+forked Accounts.
+
+```
+type AppendVecId = usize;
+
+type Fork = usize;
+
+struct AccountMap(HashMap<Fork, (AppendVecId, u64)>);
```

Comment:
Changed this to Vec<Fork, (AppendVecId, u64)>) as you had suggested.

Reply:
cool!  i can make it fit into a lock free concurrent table then :) . max length for that array should be 32

Reply:
Did you mean create an array of length 32 and replace Vec<Fork, (AppendVecId, u64)> with [(AppendVecId, u64); 32] and directly index it with fork. Would it not waste lots of memory if we did so? How would you make it lock free table, would it not require mutable reference to update it.

Reply:
@sambley yea, it would be a bit of wasted space.